Add Azure Key Vault settings source (#272)
pydantic/pydantic-settings: docs/index.md

docker service create --name pydantic-with-secrets --secret my_secret_data pydantic-app:latest
```

## Azure Key Vault

You must set two parameters:

- `url`: For example, `https://my-resource.vault.azure.net/`.

> - `credential`: If you use `DefaultAzureCredential`, in local you can execute `az login` to get your identity credentials. The identity must have a role assignement (the recommended one is `Key Vault Secrets User`), so you can have access to the secrets.

Reviewer: ```suggestion
- `credential`: If you use `DefaultAzureCredential`, in local you can execute `az login` to get your identity credentials. The identity must have a role assignment (the recommended one is `Key Vault Secrets User`), so you can access the secrets.
```

You must have the same naming convention in the field name as in the Key Vault secret name. For example, if the secret is named `SqlServerPassword`, the field name must be the same. You can use an alias too.

In Key Vault, nested models are supported with the `--` separator. For example, `SqlServer--Password`.

> JSON objects, Key Vault arrays (e.g. `MySecret--0`, `MySecret--1`) and prefixes are not supported.

Reviewer: Are you sure that we can't use prefixes? I think the new implementation supports prefixes as well.

Author: Yes, it does. Updated.

```python

> 
> ```
> import os
> from azure.identity import DefaultAzureCredential
> from pydantic import BaseModel
> from pydantic_settings import (
>     BaseSettings,
>     PydanticBaseSettingsSource
> )
> from pydantic_settings.sources import AzureKeyVaultSettingsSource
> 
> 
> class SqlServer(BaseModel):
>     password: str = Field(..., alias='Password')
> 
> class AzureKeyVaultSettings(BaseSettings):
> 
>     SqlServerPassword: str
>     sql_server_password: str = Field(..., alias='SqlServerPassword')
>     sql_server: SqlServer = Field(..., alias='SqlServer')
> 
>     @classmethod
>     def settings_customise_sources(
>         cls,
>         settings_cls: type[BaseSettings],
>         init_settings: PydanticBaseSettingsSource,
>         env_settings: PydanticBaseSettingsSource,
>         dotenv_settings: PydanticBaseSettingsSource,
>         file_secret_settings: PydanticBaseSettingsSource,
>     ) -> tuple[PydanticBaseSettingsSource, ...]:
>         return(
>             AzureKeyVaultSettingsSource(
>                 settings_cls,
>                 os.environ['AZURE_KEY_VAULT_URL'],
>                 DefaultAzureCredential()
>         )

Reviewer: ```suggestion
import os
from azure.identity import DefaultAzureCredential
from pydantic import BaseModel
from pydantic_settings import (
    BaseSettings,
    PydanticBaseSettingsSource
)
from pydantic_settings.sources import AzureKeyVaultSettingsSource

class SubModel(BaseModel):
    a: str

class AzureKeyVaultSettings(BaseSettings):
    foo: str
    bar: int
    sub: SubModel

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        az_key_vault_settings = AzureKeyVaultSettingsSource(
            settings_cls,
            os.environ['AZURE_KEY_VAULT_URL'],
            DefaultAzureCredential(),
        )
        return init_settings, env_settings, dotenv_settings, file_secret_settings, az_key_vault_settings
```

> 
> ```
> import os
> from azure.identity import DefaultAzureCredential
> from pydantic import BaseModel
> from pydantic_settings import (
>     BaseSettings,
>     PydanticBaseSettingsSource
> )
> from pydantic_settings.sources import AzureKeyVaultSettingsSource
> 
> 
> class SqlServer(BaseModel):
>     password: str = Field(..., alias='Password')
> 
> class AzureKeyVaultSettings(BaseSettings):
> 
>     SqlServerPassword: str
>     sql_server_password: str = Field(..., alias='SqlServerPassword')
>     sql_server: SqlServer = Field(..., alias='SqlServer')
> 
>     @classmethod
>     def settings_customise_sources(
>         cls,
>         settings_cls: type[BaseSettings],
>         init_settings: PydanticBaseSettingsSource,
>         env_settings: PydanticBaseSettingsSource,
>         dotenv_settings: PydanticBaseSettingsSource,
>         file_secret_settings: PydanticBaseSettingsSource,
>     ) -> tuple[PydanticBaseSettingsSource, ...]:
>         return(
>             AzureKeyVaultSettingsSource(
>                 settings_cls,
>                 os.environ['AZURE_KEY_VAULT_URL'],
>                 DefaultAzureCredential()
>         )

Reviewer: ```suggestion
import os

from azure.identity import DefaultAzureCredential
from pydantic import BaseModel

from pydantic_settings import BaseSettings, PydanticBaseSettingsSource
from pydantic_settings.sources import AzureKeyVaultSettingsSource

class SubModel(BaseModel):
    a: str

class AzureKeyVaultSettings(BaseSettings):
    foo: str
    bar: int
    sub: SubModel

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        az_key_vault_settings = AzureKeyVaultSettingsSource(
            settings_cls,
            os.environ['AZURE_KEY_VAULT_URL'],
            DefaultAzureCredential(),
        )
        return (
            init_settings,
            env_settings,
            dotenv_settings,
            file_secret_settings,
            az_key_vault_settings,
        )
```

docker service create --name pydantic-with-secrets --secret my_secret_data pydantic-app:latest
```

## Azure Key Vault

You must set two parameters:

- `url`: For example, `https://my-resource.vault.azure.net/`.
- `credential`: If you use `DefaultAzureCredential`, in local you can execute `az login` to get your identity credentials. The identity must have a role assignment (the recommended one is `Key Vault Secrets User`), so you can access the secrets.

You must have the same naming convention in the field name as in the Key Vault secret name. For example, if the secret is named `SqlServerPassword`, the field name must be the same. You can use an alias too.

In Key Vault, nested models are supported with the `--` separator. For example, `SqlServer--Password`.

Key Vault arrays (e.g. `MySecret--0`, `MySecret--1`) are not supported.

```python

> import os

Reviewer: ```suggestion
import os
from typing import Tuple, Type
```

> import os
> 
> from azure.identity import DefaultAzureCredential
> from pydantic import BaseModel
> 
> from pydantic_settings import BaseSettings, PydanticBaseSettingsSource
> from pydantic_settings.sources import AzureKeyVaultSettingsSource
> 
> 
> class SubModel(BaseModel):
>     a: str
> 
> 
> class AzureKeyVaultSettings(BaseSettings):
>     foo: str
>     bar: int
>     sub: SubModel
> 
>     @classmethod
>     def settings_customise_sources(
>         cls,
>         settings_cls: type[BaseSettings],
>         init_settings: PydanticBaseSettingsSource,
>         env_settings: PydanticBaseSettingsSource,
>         dotenv_settings: PydanticBaseSettingsSource,
>         file_secret_settings: PydanticBaseSettingsSource,
>     ) -> tuple[PydanticBaseSettingsSource, ...]:
>         az_key_vault_settings = AzureKeyVaultSettingsSource(
>             settings_cls,
>             os.environ['AZURE_KEY_VAULT_URL'],
>             DefaultAzureCredential(),
>         )
>         return (
>             init_settings,
>             env_settings,
>             dotenv_settings,
>             file_secret_settings,
>             az_key_vault_settings,
>         )

Reviewer: ```suggestion
import os
from typing import Tuple, Type

from azure.identity import DefaultAzureCredential
from pydantic import BaseModel

from pydantic_settings import BaseSettings, PydanticBaseSettingsSource
from pydantic_settings.sources import AzureKeyVaultSettingsSource

class SubModel(BaseModel):
    a: str

class AzureKeyVaultSettings(BaseSettings):
    foo: str
    bar: int
    sub: SubModel

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: Type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> Tuple[PydanticBaseSettingsSource, ...]:
        az_key_vault_settings = AzureKeyVaultSettingsSource(
            settings_cls,
            os.environ['AZURE_KEY_VAULT_URL'],
            DefaultAzureCredential(),
        )
        return (
            init_settings,
            env_settings,
            dotenv_settings,
            file_secret_settings,
            az_key_vault_settings,
        )
```

docker service create --name pydantic-with-secrets --secret my_secret_data pydantic-app:latest
```

## Azure Key Vault

You must set two parameters:

- `url`: For example, `https://my-resource.vault.azure.net/`.
- `credential`: If you use `DefaultAzureCredential`, in local you can execute `az login` to get your identity credentials. The identity must have a role assignment (the recommended one is `Key Vault Secrets User`), so you can access the secrets.

You must have the same naming convention in the field name as in the Key Vault secret name. For example, if the secret is named `SqlServerPassword`, the field name must be the same. You can use an alias too.

In Key Vault, nested models are supported with the `--` separator. For example, `SqlServer--Password`.

Key Vault arrays (e.g. `MySecret--0`, `MySecret--1`) are not supported.

> ```
> import os
> from typing import Tuple, Type
> 
> from azure.identity import DefaultAzureCredential
> from pydantic import BaseModel
> 
> from pydantic_settings import BaseSettings, PydanticBaseSettingsSource
> from pydantic_settings import AzureKeyVaultSettingsSource

Reviewer: ```suggestion
```py
import os
from typing import Tuple, Type

from azure.identity import DefaultAzureCredential
from pydantic import BaseModel

from pydantic_settings import (
    AzureKeyVaultSettingsSource,
    BaseSettings,
    PydanticBaseSettingsSource,
)
```

        )
```

### Azure Key Vault

> You must set two parameters:

Reviewer 2: we need to be clear about the limitations relative to other sources, if we can't get all KV pairs easily.

Reviewer: @AndreuCodina if we can find a way to fetch all the key/value first and build a dict, we can easily use the logic that we have in `EnvSettingsSource`. otherwise, we have to be clear and document what exactly we support and what we not support relative to other sources.

Author: Done.

- `url`: For example, `https://my-resource.vault.azure.net/`.
- `credential`: If you use `DefaultAzureCredential`, in local you can execute `az login` to get your identity credentials. The identity must have a role assignement (the recommended one is `Key Vault Secrets User`), so you can have access to the secrets.

> ```

Reviewer 2: ``````suggestion
```py
``````

Author: I can't do this because there's no Key Vault resource to test against.
